src/pudl/extract/epacems.py

```diff
@@ -146,29 +149,73 @@ def _csv_to_dataframe(self, csv_file) -> pd.DataFrame:
         return df
 
 
-def extract(epacems_years, states, ds: Datastore):
+@op(
+    out=DynamicOut(EpaCemsPartition),
+    config_schema={
+        'years': Field(Any, description='Years to process.', default_value=list(pc.WORKING_PARTITIONS['epacems']['years'])),
+        'states': Field(Any, description='States to process.', default_value=list(pc.WORKING_PARTITIONS['epacems']['states'])),
+    },
+    required_resource_keys={"pudl_engine"}
+)
+def gather_partitions(context):
+    """Gather CEMS partitions, validate settings and PUDL DB plant attributes."""
+    # Validate the partitions
+    # Could maybe replace our pydantic settings models with
+    # Dagster Types that check working partitions and tables.
+    # https://docs.dagster.io/concepts/types#defining-a-dagster-type
+    # Might be cumbersome though.
+    years = context.op_config["years"]
+    states = context.op_config["states"]
+    settings = EpaCemsSettings(years=years, states=states)
+
+    # Verify that we have a PUDL DB with plant attributes:
+    pudl_engine = context.resources.pudl_engine
+    inspector = sa.inspect(pudl_engine)
+    if "plants_eia860" not in inspector.get_table_names():
+        raise RuntimeError(
+            "No plants_eia860 available in the PUDL DB! Have you run the ETL? "
+            f"Trying to access PUDL DB: {pudl_engine}"
+        )
+
+    eia_plant_years = pd.read_sql(
+        """
+        SELECT DISTINCT strftime('%Y', report_date)
+        AS year
+        FROM plants_eia860
+        ORDER BY year ASC
+        """, pudl_engine).year.astype(int)
+
+    missing_years = list(set(settings.years) - set(eia_plant_years))
+    if missing_years:
+        logger.info(
+            f"EPA CEMS years with no EIA plant data: {missing_years} "
+            "Some timezones may be estimated based on plant state."
+        )
+
+    # Yield the partitions once everything is validated.
+    for year in settings.years:
+        for state in settings.states:
+            yield DynamicOutput(EpaCemsPartition(state=state, year=year), mapping_key=f"{state}{year}")
```

Comment:
Would we want `mapping_key` to be something longer and more guaranteed to be unique? Or is this only for differentiating the asset / output within this set of dynamically generated outputs (and they'll all start with some longer prefix? Where is that prefix / ID defined in here?)

Reply:
From the [docs](https://docs.dagster.io/_apidocs/dynamic#dagster.DynamicOutput):
> "Each DynamicOutput must have a unique mapping_key to distinguish it with it’s set."